### ISSUE TITLE
wildcard dns blocking is not working on the latest master

### ISSUE THREAD

**Issue by extremeshok** [pi-hole/pi-hole#217]
For example *.google.com:
echo 'address=/.google.com/127.0.0.1' >> /etc/dnsmasq.conf
Returns 127.0.0.1 for all google.com domains. 

However, if 
echo '.google.com' >> /etc/blacklist.txt
/usr/local/bin/gravity.sh

None of the subdomains are blocked.
##

**Comment by dschaper** [comment]
We'll have to look into that one. I'm not sure if we can do wildcard blocking, since the dnsmasq server is pulling records via a host table lookup. I don't believe dnsmasq currently is smart enough to handle wildcards, but again this will take some more research to be certain of it's capabilities.

**Comment by extremeshok** [comment]
Really not that difficult ;-P Wildcard blocking is more efficient than trying to map all the sub-domains.

Here you go, wildcard support for dnsmasq.

echo 'conf-file=/etc/dnsmasq/wildcard-blacklist.conf' >> /etc/dnsmasq.conf

echo 'address=/.google.com/127.0.0.1' >> /etc/dnsmasq/wildcard-blacklist.conf
echo 'address=/.facebook.com/127.0.0.1' >> /etc/dnsmasq/wildcard-blacklist.conf

**Comment by dschaper** [comment]
I completely agree that wildcard blocking is vastly more efficient. It would be for blacklisting/whitelisting primarily then, since we're pulling down hosts formatted lists to populate most of the blocking. Thanks for the pointers and I'll add this as an enhancement that we need to look to integrate.

**Comment by jacobsalmela** [comment]
Something to consider is that when we were using `address=//` instead of a hosts file, [performance degraded considerably when there were too many entries](http://jacobsalmela.com/huge-update-to-ad-list-that-the-pi-hole-uses/).

**Comment by dschaper** [comment]
Yeah, I remember reading that, that's why I was thinking this would be just for user whitelist/blacklist additions, not for the full Pi-Hole list.

**Comment by PromoFaux** [comment]
>echo '.google.com' >> /etc/blacklist.txt
/usr/local/bin/gravity.sh

@extremeshok Just as an aside, you can now perform nearly exactly the same thing (without rerunning gravity) by running `blacklist.sh .google.com` which just updates the blacklist, removes it from the gravity.list file, then restarts dnsmasq

**Comment by PromoFaux** [comment]
OK, this is going to take some doing. I have the logic in my head, I just need to get it out into code. At the moment I'm not going to be able to do it without breaking other features. E.G @Mcat12 's web based whitelist editor - which I think should take priority right now.

**Comment by extremeshok** [comment]
I just integrated bind9 for the wildcard, domain and record/ip rewriting.

client -> | dnsmasq -> bind9 | -> upstream dns

ie.
any dns record containing aa.bb.cc.dd will become ee.ff.gg.hh
*.google.com / google.com -> microsoft.com

__________________________________.    https://eXtremeSHOK.com     .__________________________________

**Comment by dschaper** [comment]
Ah, I see... (I've been running a second Auth DNS with dnsmasq, unbound, and nsd.)

**Comment by extremeshok** [comment]
Used to use unbound, but it fails under heavy load, also cant replace a 
specific ip in all possible domains with another ip without specifying 
the domains.

Use case is this, a ISP will hijack certain domains with a specific ip 
address and force it via their content filter/proxy.

All im doing is unhijacking the ip, and replacing it with my own 
filter/proxy.

__________________________________.    https://eXtremeSHOK.com     .__________________________________

**Comment by dschaper** [comment]
@extremeshok Have you seen #170?

**Comment by extremeshok** [comment]
#170 done ;-P

__________________________________.    https://eXtremeSHOK.com     .__________________________________

**Comment by nomoreads** [comment]
Really looking forward to wildcard domains support, as it is an ongoing issue with some pesky websites which use literally tons of subdomains to bypass blocking.

And YES it seems technically possible in DNSMasq (which is what pihole uses) https://www.leaseweb.com/labs/2013/08/wildcard-dns-ubuntu-hosts-file-using-dnsmasq/

**Comment by TylonHH** [comment]
+1

This is my *incomplete* blacklist
````
ps-14.pubnub.com
ps-862.pubnub.com
ps-827.pubnub.com
ps-906.pubnub.com
ps-121.pubnub.com
ps-566.pubnub.com
ps-565.pubnub.com
ps-332.pubnub.com
ps-494.pubnub.com
ps-803.pubnub.com
ps-446.pubnub.com
ps-107.pubnub.com
ps-43.pubnub.com
ps-630.pubnub.com
ps-731.pubnub.com
ps-152.pubnub.com
ps-393.pubnub.com
ps-479.pubnub.com
ps-687.pubnub.com
ps-177.pubnub.com
ps-317.pubnub.com
ps-985.pubnub.com
ps-459.pubnub.com
ps-150.pubnub.com
ps-564.pubnub.com
ps-748.pubnub.com
ps-861.pubnub.com
ps-678.pubnub.com
ps-723.pubnub.com
ps-124.pubnub.com
pubsub.pubnub.com
ps-306.pubnub.com
ps-87.pubnub.com
ps-936.pubnub.com
ps-361.pubnub.com
ps-198.pubnub.com
ps-120.pubnub.com
ps-662.pubnub.com
ps-84.pubnub.com
ps-7.pubnub.com
ps-528.pubnub.com
ps-990.pubnub.com
````

**Comment by PromoFaux** [comment]
Ah yeah, this one kind of fell by the wayside, but it is something we will be looking into solving. The work I had done on it will need to be looked over again as a lot of the base files have changed since then.

**Comment by TylonHH** [comment]
https://github.com/pi-hole/pi-hole/issues/672

**Comment by nomoreads** [comment]
@TylonHH Thanks for #672 

Although not officially supported, i REALLY hope the devs at Pihole will make this more user friendly.
How?

Just when i add *.bla.com to pihole's blacklist in the admin panel.
I'm not really a Linux guy as of yet so have to do this all with commands kinda makes me get lost.

I really hope the devs realize the huge value of wildcard blacklisting here, it's a huge huge time saver and REALLY stop those pesky sites that has generated literally tons of subdomains in order to bypass blocking (such as puhtml dot com, which i am blocking via hosts right now, this domain alone the .txt file is almost 30MB! way simpler with just *.puhtml dot com!)

**Comment by PromoFaux** [comment]
>hope the devs at Pihole will make this more user friendly

It's in the pipe-line! I have the logic in my head, just need to get that out as actual working code ;)

Watch this space...

**Comment by nomoreads** [comment]
@PromoFaux Would be great!

Here is my current PUHTML blocklist inside my hosts file.
Can't attache files larger than 10MB here so must download here http://www41.zippyshare.com/v/lwyVyLEL/file.html

If Notepad goes crazy you can (and should) install Notepad++ (fantastic software).

**Comment by PromoFaux** [comment]
Cool, that may come in handy when I'm testing it out! Cheers!

I'm an [ultraedit](http://www.ultraedit.com/) man myself!

**Comment by nomoreads** [comment]
Well as soon as the wildcard feature works in an easy way the file will not be needed anymore, an easy single line will solve it all :)
Basically what i did was i found puhtml is using a pattern, namely pl*.puhtml dot com (where * can be any number) so i wrote myself a script that looped 999999 times and increments the number between the "pl" and the first dot.

**Comment by PromoFaux** [comment]
Did you read [this](https://github.com/pi-hole/pi-hole/issues/672#issuecomment-247960864) comment in the above mentioned issue thread?

You can get rid of that file now if you follow the instructions there!

**Comment by nomoreads** [comment]
I'm totally new to Linux so i have to dig into the commands.
I'll check it out.
But i would appreciate getting this working via the admin panel though, i prefer GUI over CLI.

Things to keep in mind while working on this feature:

-. Is there a way to whitelist one or a few if you block a whole domain?
Will pihole first check the whitelist as soon as it gets a request?

-. What about blocking multi-subdomains?
Like one.two.three.com

*.one.two.three.com will block everything before *.one.two.three.com, thus every subdomain before .two.three.com should not be blocked. 
*.two.three.com will block everything before *.two.three.com, thus every subdomain before .three.com should not be blocked, except everything before *.two.three.com.

**Comment by St0ner1995** [comment]
just showing my support for this issue since there is no other way of doing so

**Comment by Mcat12** [comment]
@St0ner1995 you can vote here: https://discourse.pi-hole.net/t/wildcard-support-for-domain-black-white-listing/224

**Comment by PromoFaux** [comment]
To anyone following this thread, we will have an implementation of wildcard blacklisting in the next release:

https://github.com/pi-hole/pi-hole/pull/1065

**Comment by jacobsalmela** [comment]
This has been [implemented](https://discourse.pi-hole.net/tags/implemented) and should be resolved in 2.12 (#1065)
